### ISSUE TITLE
Failed type inference for multi-dimensional slicing

### ISSUE THREAD

**Issue by danielgordon10** [microsoft/pyright#2001]
I don't know if this is fixable from pyright's perspective or if it's an issue with the annotations in the individual libraries.

version 1.1.148

```python
from typing import TypeVar
import numpy as np
import torch
T = TypeVar("T", np.ndarray, torch.Tensor)
def foo1(bar: T) -> T:
    a = bar
    a[:, 10] = bar  # results in the following error
"""
Argument of type "T@foo1" cannot be assigned to parameter "val" of type "Tensor | int | float | bool" in function "__setitem__"
    Type "T@foo1" cannot be assigned to type "Tensor | int | float | bool"
      Type "ndarray* | Tensor*" cannot be assigned to type "Tensor"
      Type "ndarray* | Tensor*" cannot be assigned to type "int"
      Type "ndarray* | Tensor*" cannot be assigned to type "float"
      Type "ndarray* | Tensor*" cannot be assigned to type "bool"
      Type "ndarray* | Tensor*" cannot be assigned to type "Tensor | int | float | bool"
        "ndarray*" is incompatible with "Tensor"
        "ndarray*" is incompatible with "int" (reportGeneralTypeIssues)
"""
    return a

def foo2(bar: T) -> T:
    a = bar[:, 10]
    return a  #  Expression of type "Any | Tensor*" cannot be assigned to return type "T@foo2
```

In both cases, a and bar should be the same T type, and the slicing shouldn't change that.

**Comment by erictraut** [comment]
Thanks for the bug report. This will be fixed in the next release.

**Comment by erictraut** [comment]
This is included in pyright 1.1.151, which I just published. It will also be included in the next release of pylance.
